fdbserver/storageserver.actor.cpp

```diff
@@ -4906,6 +4967,421 @@ ACTOR Future<Void> validateRangeAgainstServers(StorageServer* data,
+				    .detail("Prefix", serverKeysPrefixFor(serverID))
+				    .detail("ResultSize", readResult.size())
+				    .detail("ReadRangeCount", readRangeCount)
+				    .detail("RetryCount", retryCount)
+				    .detail("AduitServerID", serverID);
+				readRangeCount++;
+				for (int i = 0; i < readResult.size() - 1; ++i) {
+					TraceEvent(SevVerbose, "AuditStorageShardStorageServerShardReadRangeAddToResult", serverID)
+					    .detail("ValueIsServerKeysFalse", readResult[i].value == serverKeysFalse)
+					    .detail("ServerHasKey", serverHasKey(readResult[i].value))
+					    .detail("Range", KeyRangeRef(readResult[i].key, readResult[i + 1].key))
+					    .detail("AuditServer", serverID.first());
+					if (serverHasKey(readResult[i].value)) {
+						res.push_back(Standalone(KeyRangeRef(readResult[i].key, readResult[i + 1].key)));
+					}
+				}
+				if (!readResult.more) {
+					break;
+				}
+				keyBegin = readResult.back().key;
+			}
+			break;
+		} catch (Error& e) {
+			TraceEvent(SevDebug, "AuditStorageShardStorageServerShardeError", serverID)
+			    .errorUnsuppressed(e)
+			    .detail("AuditServer", serverID.first());
+			wait(tr->onError(e));
+			readRangeCount = 0;
+			retryCount++;
+		}
+	}
+	return res;
+}
+
+ACTOR Future<Void> auditStorageStorageServerShardQ(StorageServer* data, AuditStorageRequest req) {
+	TraceEvent(SevInfo, "AuditStorageShardStorageServerShardBegin", data->thisServerID);
+	ASSERT(req.getType() == AuditType::ValidateStorageServerShard);
+	wait(data->serveAuditStorageParallelismLock.take(TaskPriority::DefaultYield));
+	state FlowLock::Releaser holder(data->serveAuditStorageParallelismLock);
+	state Version grv;
+	state Version readAtVersion;
+	state Version versionBeforeWait;
+	state std::vector<std::pair<Version, KeyRangeRef>> shardAssignments;
+	state AuditStorageState res(req.id, data->thisServerID, req.getType());
+	state std::vector<std::string> errors;
+	state Transaction tr(data->cx);
+	// We want to find out any mismatch between ownRangesSeenByServerKey and ownRangesLocalView
+	state std::vector<KeyRange> ownRangesSeenByServerKey;
+	state std::vector<KeyRange> ownRangesLocalView;
+	try {
+		if (data->version.get() == 0) {
+			TraceEvent(SevInfo, "AuditStorageShardStorageServerShardFailed", data->thisServerID)
+			    .detail("Reason", "Version is not valid")
+			    .detail("LatestVersion", data->version.get());
+			req.reply.sendError(audit_storage_failed()); // we cannot run this aduit for this time
+			return Void();
+		}
+
+		// Read local shard info and serverKeys
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		wait(store(grv, tr.getReadVersion()));
+		// When selecting which version to read, we should read at a new version
+		// As a result, any fall behind SS can eventually read the data
+		if (grv <= data->version.get()) {
+			readAtVersion = data->version.get();
+			ownRangesLocalView = getServerShardsByReadingShardInfo(data, readAtVersion);
+			tr.reset();
```

Comment:
Do we need to `reset()` here? Everything is with the same version?

Reply:
reset() is used for re-set the grv version. For the case grv <= data->version.get(), we want to read at data->version.get(). Since we have already done wait(store(grv, tr.getReadVersion())), we want to reset tr here and do tr.setVersion();